jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/analysis/controlflow/ControlFlowGraph.kt

```diff
@@ -0,0 +1,225 @@
+     */
+    val edgeMap: EdgeMap get() = _edgeMap
+    private val _edgeMap: MutableEdgeMap = mutableMapOf()
+
+    val edges: Set<Edge> get() =
+        edgeMap.flatMap { it.value }.toSet()
+
+    fun hasEdge(src: NodeIndex, dst: NodeIndex): Boolean {
+        return _edgeMap[src]?.any { it.target == dst } ?: false
+    }
+
+    fun addEdge(src: NodeIndex, dst: NodeIndex, label: EdgeLabel) {
+        _nodes.add(src)
+        _nodes.add(dst)
+        _edgeMap.updateInplace(src, default = mutableSetOf()) {
+            add(Edge(src, dst, label))
+        }
+    }
+}
+
+/**
+ * Checks if the given opcode corresponds to a recognized JVM conditional jump instruction.
+ */
+internal fun isRecognizedIfJumpOpcode(opcode: Int): Boolean = when (opcode) {
+    Opcodes.IFEQ, Opcodes.IFNE, Opcodes.IFLT, Opcodes.IFGE, Opcodes.IFGT, Opcodes.IFLE,
+    Opcodes.IF_ICMPEQ, Opcodes.IF_ICMPNE, Opcodes.IF_ICMPLT, Opcodes.IF_ICMPGE, Opcodes.IF_ICMPGT, Opcodes.IF_ICMPLE,
+    Opcodes.IFNULL, Opcodes.IFNONNULL,
+    Opcodes.IF_ACMPEQ, Opcodes.IF_ACMPNE
+         -> true
+    else -> false
+}
+
+/**
+ * Checks if the given opcode corresponds to a recognized JVM jump/switch instruction.
+ */
+internal fun isRecognizedJumpOpcode(opcode: Int): Boolean = when (opcode) {
+    // Goto
+    Opcodes.GOTO,
+    // Single-operand zero comparisons
+    Opcodes.IFEQ, Opcodes.IFNE, Opcodes.IFLT, Opcodes.IFGE, Opcodes.IFGT, Opcodes.IFLE,
+    // Single-operand null comparisons
+    Opcodes.IFNULL, Opcodes.IFNONNULL,
+    // Two-operand numeric comparisons
+    Opcodes.IF_ICMPEQ, Opcodes.IF_ICMPNE, Opcodes.IF_ICMPLT, Opcodes.IF_ICMPGE, Opcodes.IF_ICMPGT, Opcodes.IF_ICMPLE,
+    // Two-operand object reference comparisons
+    Opcodes.IF_ACMPEQ, Opcodes.IF_ACMPNE,
+    // Switches
+    Opcodes.TABLESWITCH, Opcodes.LOOKUPSWITCH,
+         -> true
+
+    Opcodes.JSR, Opcodes.RET -> {
+        Logger.warn {
+            "Deprecated JSR/RET instructions are encountered during control-flow graph construction."
+        }
+        true
+    }
+
+    else -> false
+}
+
+internal fun Collection<Edge>.prettyPrint(): String {
+    val sb = StringBuilder("")
+    val edges = sortedEdges()
+    for ((id, e) in edges.withIndex()) {
+        val label = e.label.prettyPrint().takeIf { it.isNotEmpty() }
+        sb.append("B${e.source} -> B${e.target}")
+        sb.append(label?.let { " : $it" }.orEmpty())
+        if (id != edges.lastIndex) sb.appendLine()
+    }
+    return sb.toString()
+}
+
+private fun Collection<Edge>.sortedEdges(): List<Edge> = toMutableList().apply {
+    fun labelSortKey(label: EdgeLabel): String = when (label) {
+        is EdgeLabel.FallThrough -> "0"
+        is EdgeLabel.Jump        -> "1:${Printer.OPCODES[label.opcode]}"
+        is EdgeLabel.Exception   -> "2:${label.caughtTypeName ?: "*"}"
+    }
+
+    sortWith(compareBy({ it.source }, { it.target }, { labelSortKey(it.label) }))
+}
+
+private fun EdgeLabel.prettyPrint(): String = when (this) {
```

Comment:
Note: I've extracted it here as well

Pretty print is the name used inside a CFGPrinter class, here some other name should be used

Reply:
I have overriden `toString` methods of these edge labels instead